tests/interlocked_int/interlocked_int.c

```diff
@@ -0,0 +1,855 @@
+
+    ///assert
+    ASSERT_ARE_EQUAL(int32_t, INT32_MIN + 1, addend);
+    ASSERT_ARE_EQUAL(int32_t, INT32_MIN + 1, return_val);
+
+}
+
+/*
+Tests_SRS_INTERLOCKED_43_022 [ interlocked_increment_16 shall atomically increment (increase by one) the 16-bit variable *addend.]
+Tests_SRS_INTERLOCKED_43_053: [ interlocked_increment_16 shall return the incremented value. ]
+*/
+TEST_FUNCTION(interlocked_increment_16_upper_bound)
+{
+    ///arrange
+    volatile int16_t addend = INT16_MAX - 1;
+
+    ///act
+    int16_t return_val = interlocked_increment_16(&addend);
+
+    ///assert
+    ASSERT_ARE_EQUAL(int16_t, INT16_MAX, addend);
+    ASSERT_ARE_EQUAL(int16_t, INT16_MAX, return_val);
+}
+/*
+Tests_SRS_INTERLOCKED_43_022 [ interlocked_increment_16 shall atomically increment (increase by one) the 16-bit variable *addend.]
+Tests_SRS_INTERLOCKED_43_053: [ interlocked_increment_16 shall return the incremented value. ]
+*/
+TEST_FUNCTION(interlocked_increment_16_lower_bound)
+{
+    ///arrange
+    volatile int16_t addend = INT16_MIN;
+
+    ///act
+    int16_t return_val = interlocked_increment_16(&addend);
+
+    ///assert
+    ASSERT_ARE_EQUAL(int16_t, INT16_MIN + 1, addend);
+    ASSERT_ARE_EQUAL(int16_t, INT16_MIN + 1, return_val);
+}
+
+/*
+Tests_SRS_INTERLOCKED_43_023 [ interlocked_increment_64 shall atomically increment (increase by one) the 64-bit variable *addend.]
+Tests_SRS_INTERLOCKED_43_054: [ interlocked_increment_64 shall return the incremented value. ]
+*/
+TEST_FUNCTION(interlocked_increment_64_upper_bound)
+{
+    ///arrange
+    volatile int64_t addend = INT64_MAX - 1;
+
+    ///act
+    int64_t return_val = interlocked_increment_64(&addend);
+
+    ///assert
+    ASSERT_ARE_EQUAL(int64_t, INT64_MAX, addend);
+    ASSERT_ARE_EQUAL(int64_t, INT64_MAX, return_val);
+}
+/*
+Tests_SRS_INTERLOCKED_43_023 [ interlocked_increment_64 shall atomically increment (increase by one) the 64-bit variable *addend.]
+Tests_SRS_INTERLOCKED_43_054: [ interlocked_increment_64 shall return the incremented value. ]
+*/
+TEST_FUNCTION(interlocked_increment_64_lower_bound)
+{
+    ///arrange
+    volatile int64_t addend = INT64_MIN;
+
+    ///act
+    int64_t return_val = interlocked_increment_64(&addend);
+
+    ///assert
+    ASSERT_ARE_EQUAL(int64_t, INT64_MIN + 1, addend);
+    ASSERT_ARE_EQUAL(int64_t, INT64_MIN + 1, return_val);
+
+}
+
+/*
+Tests_SRS_INTERLOCKED_43_024 [ interlocked_or shall perform an atomic bitwise OR operation on the 32-bit integers *destination and value and store the result in destination.]
+Tests_SRS_INTERLOCKED_43_055: [ interlocked_or shall return the initial value of *destination. ]
+*/
+
+TEST_FUNCTION(interlocked_or_does_bitwise_or)
+{
+    ///arrange
+    volatile uint32_t destination = 0xF0F0F0F0;
+    uint32_t value = 0x0F0F0F0F;
+
+    ///act
+    uint32_t return_val = (uint32_t)interlocked_or((volatile int32_t*)&destination, (int32_t)value);
+
+    ///assert
+    ASSERT_ARE_EQUAL(uint32_t, -1, destination);
```

Comment:
>-1 [](start = 31, length = 2)

-1 and uint32_t (unsigned) don't play nice together. Another value is proper here.

Reply:
Seems like a case of "here and everywhere" when comparing unsigned values and -1.

---
In reply to: [427738341](https://github.com/Azure/azure-c-util/pull/43#discussion_r427738341) [](ancestors = 427738341)

Reply:
Matt suggested using unsigned integers because when I tried using signed integers, I wasn't able to put n bits in an n-bit signed integer. So I declare them as unsigned and then typecast to signed when passing to the function. I can change the type in the ASSERT to make it clear that I am comparing two signed values.

---
In reply to: [427738515](https://github.com/Azure/azure-c-util/pull/43#discussion_r427738515) [](ancestors = 427738515,427738341)

Reply:
Parth, I think you just want to change it to:

```
ASSERT_ARE_EQUAL(uint32_t, 0xFFFFFFFF, destination);
```

Right now you are using a signed value and saying the type is unsigned.

---
In reply to: [428355088](https://github.com/Azure/azure-c-util/pull/43#discussion_r428355088) [](ancestors = 428355088,427738515,427738341)